ext/random/randomizer.c

```diff
@@ -88,6 +88,118 @@ PHP_METHOD(Random_Randomizer, __construct)
 }
 /* }}} */
 
+/* {{{ Generate a float in [0, 1) */
+PHP_METHOD(Random_Randomizer, nextFloat)
+{
+	php_random_randomizer *randomizer = Z_RANDOM_RANDOMIZER_P(ZEND_THIS);
+	uint64_t result;
+	size_t total_size;
+
+	ZEND_PARSE_PARAMETERS_NONE();
+
+#ifndef __STDC_IEC_559__
+	zend_throw_exception(random_ce_Random_RandomException, "The nextFloat() method requires the underlying 'double' representation to be IEEE-754.", 0);
+	RETURN_THROWS();
+#endif
+
+	result = 0;
+	total_size = 0;
+	do {
+		uint64_t r = randomizer->algo->generate(randomizer->status);
+		result = result | (r << (total_size * 8));
+		total_size += randomizer->status->last_generated_size;
+		if (EG(exception)) {
+			RETURN_THROWS();
+		}
+	} while (total_size < sizeof(uint64_t));
+
+	/* A double has 53 bits of precision, thus we must not
+	 * use the full 64 bits of the uint64_t, because we would
+	 * introduce a bias / rounding error.
+	 */
+	ZEND_ASSERT(DBL_MANT_DIG == 53);
+	const double step_size = 1.0 / (1ULL << 53);
+
+	/* Use the upper 53 bits, because some engine's lower bits
+	 * are of lower quality.
+	 */
+	result = (result >> 11);
+
+	RETURN_DOUBLE(step_size * result);
+}
+/* }}} */
+
+static double getFloat_gamma_low(double x)
+{
+	return x - nextdown(x);
+}
+
+static double getFloat_gamma_high(double x)
+{
+	return nextup(x) - x;
+}
+
+static double getFloat_gamma(double x, double y)
+{
+	return (fabs(x) > fabs(y)) ? getFloat_gamma_high(x) : getFloat_gamma_low(y);
+}
+
+static uint64_t getFloat_ceilint(double a, double b, double g)
+{
+	double s = b / g - a / g;
+	double e;
+
+	if (fabs(a) <= fabs(b)) {
+		e = -a / g - (s - b / g);
+	} else {
+		e = b / g - (s + a / g);
+	}
+
+	double si = ceil(s);
+
+	return (s != si) ? (uint64_t)si : (uint64_t)si + (e > 0);
+}
+
+/* {{{ Generates a random float within [min, max).
+ *
+ * The algorithm used is the γ-section algorithm as published in:
+ *
+ * Drawing Random Floating-Point Numbers from an Interval. Frédéric
+ * Goualard, ACM Trans. Model. Comput. Simul., 32:3, 2022.
+ * https://doi.org/10.1145/3503512
+ */
+PHP_METHOD(Random_Randomizer, getFloat)
+{
```

Comment:
This seems like it may be useful for extensions -- could you extract it into a `ZEND_API` and delegate to it from this method? Or at least add a todo item saying that if the RFC passes we should do that?

Reply:
What part of this function are you interested in / what would you expect the API to look like? `PHP_API random_get_float(const php_random_algo *algo, php_random_status *status, double min, double max)`?

Reply:
@morrisonlevi I believe this is now done with gammasection.c, can you take a look if this is what you had in mind?

Reply:
I have not reviewed the API, but yes, this is what I had in mind. Thank you!